Detectors/EMCAL/base/src/ClusterFactory.cxx

```diff
@@ -551,24 +555,145 @@ void ClusterFactory<InputType>::evalElipsAxis(gsl::span<const int> inputsIndices
+  return std::make_tuple(mid, energy, cellAmp, shared);
+}
+
+///
+/// Look to cell neighbourhood and reject if it seems exotic
+//____________________________________________________________________________
+template <class InputType>
+bool ClusterFactory<InputType>::isExoticCell(short towerId, float ecell) const
+{
+  if (ecell < mExoticCellMinAmplitude) {
+    return false; // do not reject low energy cells
+  }
+
+  float eCross = getECross(towerId, ecell);
+
+  if (1 - eCross / ecell > mExoticCellFraction) {
+    LOG(debug) << Form("EXOTIC CELL id %d, eCell %f, eCross %f, 1-eCross/eCell %f",
+                       towerId, ecell, eCross, 1 - eCross / ecell);
+    return true;
+  }
+
+  return false;
+}
+
+///
+///  Calculate the energy in the cross around the energy of a given cell.
+//____________________________________________________________________________
+template <class InputType>
+float ClusterFactory<InputType>::getECross(short towerId, float energy) const
+{
+  auto [iSM, iMod, iIphi, iIeta] = mGeomPtr->GetCellIndex(towerId);
+  auto [iphi, ieta] = mGeomPtr->GetCellPhiEtaIndexInSModule(iSM, iMod, iIphi, iIeta);
+
+  // Get close cells index, energy and time, not in corners
+
+  short towerId1 = -1;
+  short towerId2 = -1;
+
+  if (iphi < o2::emcal::EMCAL_ROWS - 1)
+    towerId1 = mGeomPtr->GetAbsCellIdFromCellIndexes(iSM, iphi + 1, ieta);
+  if (iphi > 0)
+    towerId2 = mGeomPtr->GetAbsCellIdFromCellIndexes(iSM, iphi - 1, ieta);
+
+  // In case of cell in eta = 0 border, depending on SM shift the cross cell index
+
+  short towerId3 = -1;
+  short towerId4 = -1;
+
+  if (ieta == o2::emcal::EMCAL_COLS - 1 && !(iSM % 2)) {
+    towerId3 = mGeomPtr->GetAbsCellIdFromCellIndexes(iSM + 1, iphi, 0);
+    towerId4 = mGeomPtr->GetAbsCellIdFromCellIndexes(iSM, iphi, ieta - 1);
+  } else if (ieta == 0 && iSM % 2) {
+    towerId3 = mGeomPtr->GetAbsCellIdFromCellIndexes(iSM, iphi, ieta + 1);
+    towerId4 = mGeomPtr->GetAbsCellIdFromCellIndexes(iSM - 1, iphi, o2::emcal::EMCAL_COLS - 1);
+  } else {
+    if (ieta < o2::emcal::EMCAL_COLS - 1)
+      towerId3 = mGeomPtr->GetAbsCellIdFromCellIndexes(iSM, iphi, ieta + 1);
+    if (ieta > 0)
+      towerId4 = mGeomPtr->GetAbsCellIdFromCellIndexes(iSM, iphi, ieta - 1);
+  }
+
+  LOG(debug) << Form("iSM %d, towerId %d, a %d, b %d, c %d e %d", iSM, towerId, towerId1, towerId2, towerId3, towerId4);
+
+  float ecell1 = 0, ecell2 = 0, ecell3 = 0, ecell4 = 0;
+
+  short iTowerId = -1;
+  for (auto& cellIndex : mCellsIndices) {
+    iTowerId = mInputsContainer[cellIndex].getTower();
+    if (towerId1 == iTowerId) {
+      ecell1 = mInputsContainer[cellIndex].getEnergy();
+    }
+    if (towerId2 == iTowerId) {
+      ecell1 = mInputsContainer[cellIndex].getEnergy();
+    }
+    if (towerId3 == iTowerId) {
+      ecell1 = mInputsContainer[cellIndex].getEnergy();
+    }
+    if (towerId4 == iTowerId) {
+      ecell1 = mInputsContainer[cellIndex].getEnergy();
+    }
```

Comment:
I think this implementation is not correct. mCellIndicies are the cells indices only of all clustered cells, while the Fcross cut takes the neightboring cells of the leading cell in order to calculate the Fcross, no matter whehter a cell is clustered. You have to loop over all entries in mInputContainer and check whether the tower index of the cell is among one of the 4 neighbors.

Reply:
Also it should not always be ecell1 that is getting set. :sweat_smile: